modules/transports/core/nhttp/src/main/java/org/apache/synapse/transport/passthru/TargetRequest.java

```diff
@@ -139,108 +137,135 @@ public void start(NHttpClientConnection conn) throws IOException, HttpException
-			OMOutputFormat format = PassThroughTransportUtils.getOMOutputFormat(requestMsgCtx);
-			if (formatter != null && format != null) {
-				URL _url = formatter.getTargetAddress(requestMsgCtx, format, url);
-				if (_url != null && !_url.toString().isEmpty()) {
-					if (requestMsgCtx.getProperty(NhttpConstants.POST_TO_URI) != null
-							&& Boolean.TRUE.toString().equals(requestMsgCtx.getProperty(NhttpConstants.POST_TO_URI))) {
-						path = _url.toString();
-					} else {
-						path = _url.getPath()
-								+ ((_url.getQuery() != null && !_url.getQuery().isEmpty())
-										? ("?" + _url.getQuery())
-										: "");
-					}
-
-				}
-				headers.remove(HTTP.CONTENT_TYPE);
-			}
-		}        
-        
+        MessageContext requestMsgCtx = TargetContext.get(conn).getRequestMsgCtx();
         
-		Object o = requestMsgCtx.getProperty(MessageContext.TRANSPORT_HEADERS);
-		if (o != null && o instanceof TreeMap) {
-			Map _headers = (Map) o;
-			String trpContentType = (String) _headers.get(HTTP.CONTENT_TYPE);
-			if (trpContentType != null && !trpContentType.equals("")) {
-				if (!TargetRequestFactory.isMultipartContent(trpContentType)) {
-					addHeader(HTTP.CONTENT_TYPE, trpContentType);
-				}
-			}
-
-		}
+        HttpRequest request = 
+                new BasicHttpEntityEnclosingRequest(method, path, version != null ? version : HttpVersion.HTTP_1_1);
+
+        BasicHttpEntity entity = new BasicHttpEntity();
         
+        ((BasicHttpEntityEnclosingRequest) request).setEntity(entity);
         
-  
-                                                            
-        if (hasEntityBody) {
-            request = new BasicHttpEntityEnclosingRequest(method, path,
-                    version != null ? version : HttpVersion.HTTP_1_1);
-
-            BasicHttpEntity entity = new BasicHttpEntity();
-
-
-            boolean forceContentLength = requestMsgCtx.isPropertyTrue(
-                                                                   NhttpConstants.FORCE_HTTP_CONTENT_LENGTH);
-            boolean forceContentLengthCopy = requestMsgCtx.isPropertyTrue(
-                                                                   PassThroughConstants.COPY_CONTENT_LENGTH_FROM_INCOMING);
-                                
-            if (forceContentLength) {
-                entity.setChunked(false);
-                if (forceContentLengthCopy && contentLength > 0) {
-                    entity.setContentLength(contentLength);
-                } 
-            }else{
-             if (contentLength != -1) {
+        boolean forceContentLength = 
+                requestMsgCtx.isPropertyTrue(NhttpConstants.FORCE_HTTP_CONTENT_LENGTH);
+        boolean forceContentLengthCopy = 
+                requestMsgCtx.isPropertyTrue(PassThroughConstants.COPY_CONTENT_LENGTH_FROM_INCOMING);
+                            
+        if (forceContentLength) {
+            entity.setChunked(false);
+            if (forceContentLengthCopy && contentLength != -1) {
```

Comment:
What if contentLength is 0?

Reply:
If the target API forces a Content-Length to be sent, the API will expect Content-Length even for POST requests without a Body.(Content-Length: 0). The main purpose of this pull request is that.

#841 solves the java.lang.ClassCastException, but when chunking is disabled Content-Type is not set.